File: SkiCameraAndroid/shared/src/commonMain/kotlin/com/vcnt/skicamera/shared/SmoothingFilter.kt
package com.vcnt.skicamera.shared

/**
 * Exponential Moving Average (EMA) filter.
 * [alpha] determines Smoothing vs. Lag.
 */
class SmoothingFilter(var alpha: Double = 0.5) {
    private var prevValue: Double? = null

    fun filter(rawValue: Double): Double {
        val prev = prevValue
        return if (prev == null) {
            prevValue = rawValue
            rawValue
        } else {
            val newValue = (alpha * rawValue) + ((1.0 - alpha) * prev)
            prevValue = newValue
            newValue
        }
    }

    fun currentValue(): Double? = prevValue

    fun reset() {
        prevValue = null
    }
}

File: SkiCameraIOS/SkiCameraIOS/Camera/Logic/SmoothingFilter.swift
import Foundation

class SmoothingFilter {
    var alpha: Double
    private var value: Double?
    
    init(alpha: Double) {
        self.alpha = alpha
    }
    
    func filter(_ input: Double) -> Double {
        guard let current = value else {
            value = input
            return input
        }
        let newValue = (alpha * input) + ((1.0 - alpha) * current)
        value = newValue
        return newValue
    }
    
    func reset() {
        value = nil
    }
}
